ENH: Initial implementation of `apply` function

adbugger opened the pull request:
Used the existing `as_dcm` function for application.

adbugger committed:
ENH: Initial implementation of `apply` function

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -657,3 +658,32 @@ def __mul__(self, other):
             result = result[0]
         return self.__class__(result, normalized=True)
+
+    def apply(self, points):
```
I'm not sure if I like name `points` here. The conceptual problem with it is that rotation of points makes sense around some origin and we might want also apply translation to points. That is the position of "points" relative to the frame origin is important.

But since this class represents only rotations, we are better to apply the rotation to "free vectors", that is which starting point is irrelevant.

So I suggest to use either `v` or `vectors`.

nmayorov replied:
What about `inverse=False` flag?

Implementing would be really simple --- adding `.T` attribute to the DCM.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -657,3 +658,32 @@ def __mul__(self, other):
+
+        Rotates `points` by the rotation(s) represented in the object. In terms
+        of DCMs, this application is the same as `self.as_dcm().dot(points)`.
+        For an object containing `N` rotations applied on `P` points, returns
```
Are you sure this semantics is preferable? My initial thinking was that we would use the same semantics as with rotation composition. That is in "single rotation, multiple vectors", "multiple rotations, single vector" or "stack of rotations, stack of vectors of the same size".

In my practice, I only needed this "stack semantics". That is we have a time series of rotations and vectors in one frame and want to transform all of them to another. I'm not sure that this "dot" semantics would mean.

@larsoner, @adbugger  what do you think?

adbugger replied:
+0.75 for the `stack` semantics. Although I don't think `dot` semantics are "wrong" per se, it makes more sense to stick to the convention, and it is backed by practical application.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -657,3 +658,32 @@ def __mul__(self, other):
+        """
+        points = np.asarray(points)
+        if points.ndim > 2 or points.shape[-1] != 3:
+            raise ValueError("Expected input of shape (3,) or (N, 3), "
```
In this formulation `(P, 3)`. If we change to "stack semantics", then `(N, 3)` everywhere.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -657,3 +658,32 @@ def __mul__(self, other):
+    def apply(self, points):
+        """Apply this rotation on a set of points.
+
+        Rotates `points` by the rotation(s) represented in the object. In terms
```
I think this is an appropriate place to explain "rotation-projection duality". That is we can either think of it as projection of a vector components from one frame to another, or physical rotation of a vector as if it is glued to the original frame as it rotates.

nmayorov reviewed (commented):
There is a quite significant question of "stack vs dot" semantics. I'm quite firmly inclined to implement "stack" semantics (opposite to what you implemented).

adbugger committed:
MAINT/DOC: Explain rotation-projection duality, add inverse flag

adbugger commented:
Added `inverse` flag, explained the rotation-projection duality, and changed `points` to `vectors`. While I also think `stack` semantics are preferable, I haven't changed those semantics for now.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -657,3 +658,47 @@ def __mul__(self, other):
+    def apply(self, vectors, inverse=False):
+        """Apply this rotation on a set of vectors.
+
+        There are 2 ways to visualize a rotation of vectors:
```
I suggest to write something different (this is just my try):
```
If the original frame rotates to the final frame by this rotation, 
then its application to a vector can be seen in two ways:

- As a projection of vector components expressed in the final frame to the original frame.
- As a physical rotation of a vector being glued to the original frame as it rotates. 
In this case the vector components are expressed in the original frame before and after the rotation.
```

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -657,3 +658,47 @@ def __mul__(self, other):
+            - It can also be thought of as a physical rotation of a vector
+              being glued to the original frame as it rotates.
+
+        Rotates `vectors` by the rotation(s) represented in the object. In
```
I think this should go before the previous explanation.

nmayorov reviewed (commented):
Prepare some tests.

adbugger committed:
DOC: Better explanation of rotation-projection duality

adbugger commented:
Will prepare tests after semantics have been finalised.

nmayorov commented:
@larsoner this PR is in pending state, because we want to know your opinion.

Here is the question repeated:

> Are you sure this semantics is preferable? My initial thinking was that we would use the same semantics as with rotation composition. That is in "single rotation, multiple vectors", "multiple rotations, single vector" or "stack of rotations, stack of vectors of the same size".

> In my practice, I only needed this "stack semantics". That is we have a time series of rotations and vectors in one frame and want to transform all of them to another. I'm not sure that this "dot" semantics would mean.

> @larsoner, @adbugger what do you think?

adbugger commented:
> +0.75 for the stack semantics. Although I don't think dot semantics are "wrong" per se, it makes more sense to stick to the convention, and it is backed by practical application.

larsoner commented on scipy/spatial/transform/rotation.py:
```diff
@@ -657,3 +658,48 @@ def __mul__(self, other):
+            result = np.einsum('...ij,...kj->...ki', self.as_dcm(), vectors)
+
+        # For single rotation, the output is 2D, so add another axis
+        return result[None, :] if self._single else result
```
This seems inconsistent to me. I would expect the output to be `(P, 3)` if `self._single`, otherwise this output gets an extra `(1,)` singleton dimension added whereas most other outputs (including `self.as_dcm()`?) will not, right?

larsoner reviewed (commented):
So in the current iteration, given `N` notations and `P` vectors (shape `(P, 3)`) you end up with `(N, P, 3)` (an all-to-all sort of operation).

What is the API change (if any) and proposed documentation change for the "stack" semantics here? That if you have `N==P` then there should be an option to have it return vectors of shape `(P, 3)` rather than `(N, P, 3)`?

Incidentally, this is something that I think could be addressed by standard NumPy broadcasting rules if we were to adopt them. If you had rotations of shape `(N, 1)` and `P` vectors then your output would be of shape `(N, P, 3)`; if you had rotations of shape `(N,)` then `apply` with `P` vectors would only work if `P == 1`, `P == N`, or `N == 1`.

BTW this also makes me realize that this function should also take arrays of shape `(3,)` and return an output of shape `(3,) if self._single else (N, 3)`, which I'm not sure it currently does.

adbugger commented:
@larsoner you're right. The "stack" semantics suggested are as follows:
- for `N==P` return `(P, 3)`. `ith` rotation applied on `ith` point.
- for `N==1` return `(P, 3)`, unless
    - If `self.single == True` and points were specified as `(3,)`, then return `(3,)`
- for `P==1` return `(N, 3)`, unless
    - If `self.single == True` and points were specified as `(3,)`, then return `(3,)`
- otherwise we raise an error. The "all-to-all" behaviour will be removed completely.

@nmayorov do these specifications look alright to you?

nmayorov commented:
Yes, let's settle on this and finish this PR. Also include indexing/copying as a separate pull request and prepare a pull request into scipy.

And rebase commits in slerp pull request on the updated rotation branch.

adbugger committed:
ENH: dot semantics for apply

adbugger committed:
DOC: Shorter explanation of output cases

adbugger committed:
TST: Initial tests for rotation application

adbugger commented:
Apologies this one took so long. Documenting and implementing all the possible cases was a little tricky but I think I have found an implementation that is relatively straightforward. Also included tests to check for all possible input shape cases.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -657,3 +658,70 @@ def __mul__(self, other):
+              to the original frame.
+            - As the physical rotation of a vector being glued to the original
+              frame as it rotates. In this case the vector components are
+              expressed in the original frame before and after rotation.
```
"the rotation".

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -657,3 +658,70 @@ def __mul__(self, other):
+
+        Parameters
+        ----------
+        vectors : array_like, shape (3,) or (P, 3)
```
Should it be (N, 3)?

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -657,3 +658,70 @@ def __mul__(self, other):
+        Returns a `numpy.ndarray` of shape `(3,)` if object contains a single
+        rotation (as opposed to a stack with a single rotation) and a single
+        vector is specified with shape `(3,)`. In all other cases, the returned
+        array has shape `(N, 3)`.
```
Should you say that `N` is either the number of rotations or number of vectors?

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -657,3 +658,70 @@ def __mul__(self, other):
+            dcm = dcm[None, :, :]
+
+        n_vectors = vectors.shape[0]
+        n_rotations = self._quat.shape[0]
```
To be changed with `self.n`?

adbugger replied:
I've chosen to put this property in the `rotation_index_copy` branch. The apply function already uses `self.n` in PR #17.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -657,3 +658,70 @@ def __mul__(self, other):
+        if self._single and single_vector:
+            return result[0]
+        else:
+            return result
```
Looks really good. Well done!

nmayorov commented on scipy/spatial/transform/tests/test_rotation.py:
```diff
@@ -619,3 +619,86 @@ def test_inv_single_rotation():
+        [1, 0, 0],
+        [0, 0, 1]
+    ])
+    r_2d = Rotation.from_dcm([[
```
Can we use `np.atleast_3d` or something?

adbugger replied:
So `np.atleast_3d` adds dimensions to the end. It takes us from `(3, 3) -> (3, 3, 1)` whereas we require `(3, 3) -> (1, 3, 3)`.
The [documentation](https://docs.scipy.org/doc/numpy-1.14.0/reference/generated/numpy.atleast_3d.html) for this function does not show a way to achieve this so I'm using `np.expand_dims` instead.

nmayorov commented on scipy/spatial/transform/tests/test_rotation.py:
```diff
@@ -619,3 +619,86 @@ def test_inv_single_rotation():
+    ]])
+
+    v_1d = np.array([1, 2, 3])
+    v_2d = np.array([[1, 2, 3]])
```
Same here, but really minor.

adbugger replied:
`np.atleast_2d` has the same issue as above, using `np.expand_dims`.

nmayorov reviewed (approved):
Looks great. The small thing missing is to add at least one tests with `inverse=False`. 

Add this and merge.

adbugger committed:
DOC: Small changes to docstrings.

adbugger committed:
TST: Tests use np.expand_dims function

adbugger committed:
TST: Add test with inverse=True
